k2/csrc/array_ops.cu

```diff
@@ -373,4 +374,77 @@ Array1<int32_t> RowSplitsToSizes(const Array1<int32_t> &row_splits) {
   return sizes;
 }
 
+
+//  This is modified from RowSplitsToRowIdsKernel.
+//  When we invoke this we make a big enough grid that there doesn't have to
+//  be a loop over rows, i.e. (gridDim.x * blockDim.x) / threads_per_row >=
+//  num_rows
+__global__ void SizesToMergeMapKernel(int32_t num_rows,
+                                        int32_t threads_per_row,
+                                        const int32_t *row_splits,
+                                        int32_t num_elems, int32_t *row_ids) {
+  int32_t thread = blockIdx.x * blockDim.x + threadIdx.x,
+          num_threads = gridDim.x * blockDim.x, row = thread / threads_per_row,
+          thread_this_row = thread % threads_per_row;
+
+  if (row >= num_rows) return;
+  K2_CHECK_GE(num_threads / threads_per_row, num_rows);
+
+  int32_t this_row_split = row_splits[row],
+          next_row_split = row_splits[row + 1],
+          row_length = next_row_split - this_row_split;
+
+#pragma unroll (4)
+  for (; thread_this_row < row_length; thread_this_row += threads_per_row)
+    row_ids[this_row_split + thread_this_row] = row;
+}
+
+
+Array1<int32_t> SizesToMergeMap(ContextPtr c, const std::vector<int32_t> &sizes) {
+  int32_t num_srcs = sizes.size();
+
+  ContextPtr cpu_context = GetCpuContext();
+  Array1<int32_t> row_splits_cpu(cpu_context, num_srcs + 1);
+  int32_t *row_splits_cpu_data = row_splits_cpu.Data();
+  int32_t tot_size = 0;
+  row_splits_cpu_data[0] = 0;
+  for (int32_t i = 0; i < num_srcs; i++) {
+    tot_size += sizes[i];
+    row_splits_cpu_data[i+1] = tot_size;
```

Comment:
just a question, seems we are doing `exlcusiveSum` here? Can we call `ExclusiveSum` here ( it works for CPU as well), I mean, something like below:
```
Array1<int32_t> array_sizes(cpu, sizes);
Exclusive(array_sizes, &row_splits_cpu); // note this is the version that dst.Dim() == row_splits_cpu.Dim() = src.Dim() +1
```

Reply:
mm yeah in principle, but would have unnecessary copy... would prefer to leave it.